client/lib/user/user.d.ts

```diff
@@ -0,0 +1,55 @@
+/**
+ * External Dependencies
+ */
+import { EventEmitter } from 'events';
+
+export class User extends EventEmitter {
+	initialize: () => Promise< void >;
+	clearStoreIfChanged: () => void;
+	get: () => UserData;
+	fetch: () => Promise< void >;
+	handleFetchFailure: () => void;
+	handleFetchSuccess: () => void;
+	getLanguage: () => string;
+	getAvatarUrl: () => string;
+	clear: () => void;
+	sendVerificationEmail: () => Promise< void >;
+	set: () => boolean;
+	decrementSiteCount: () => void;
+	incrementSiteCount: () => void;
+	verificationPollerCallback: () => void;
+	checkVerification: () => void;
+	signalVerification: () => void;
+	dispatchToken: () => void;
+}
+
+export class UserData {
+	ID: string;
+	display_name?: string;
+	username?: string;
+	avatar_URL?: string;
+	site_count?: string;
+	visible_site_count?: string;
+	date?: string;
+	has_unseen_notes?: string;
+	newest_note_type?: string;
+	phone_account?: string;
+	email?: string;
+	email_verified?: string;
+	is_valid_google_apps_country?: string;
+	user_ip_country_code?: string;
+	logout_URL?: string;
+	primary_blog?: string;
+	primary_blog_is_jetpack?: string;
+	primary_blog_url?: string;
+	meta?: string;
+	is_new_reader?: string;
+	social_login_connections?: string;
+	abtests?: string;
+	description?: string;
+	user_URL?: string;
+	primarySiteSlug?: string;
+	localeSlug?: string;
+	localeVariant?: string;
+	isRTL?: string;
+}
```

Comment:
A couple notes overall on this approach:

Is there a particular reason to create this as a `.d.ts` file instead of simply making `user.ts` or `user/types.ts`? I don't think there's anything wrong with the declaration file other than it involves or implies a certain amount of manual work that may uncouple from the code.

Is there a reason for using a `class` declaration here over a simpler `type`? As far as I could tell from the user code we're not actually dealing with a class but an object of some of many possible properties. What I'm going to share about `Partial` can apply to the `class` declaration as well but it's awkward in my reasoning to apply it there, especially since I don't think this is actually a class we're wanting to model.

Have you heard about [`Partial`](http://www.typescriptlang.org/docs/handbook/utility-types.html#partialt)? I think you might like it.

Consider this construction; I'd love to talk about it with you and hear your thoughts.

```ts
export type UserData = { ID: string } & Partial<OptionalUserData>;

export type OptionalUserData = {
 	display_name: string;
 	username: string;
 	avatar_URL: string;
 	site_count: string;
 	visible_site_count: string;
 	date: string;
 	has_unseen_notes: string;
 	newest_note_type: string;
 	phone_account: string;
 	email: string;
 	email_verified: string;
 	is_valid_google_apps_country: string;
 	user_ip_country_code: string;
 	logout_URL: string;
 	primary_blog: string;
 	primary_blog_is_jetpack: string;
 	primary_blog_url: string;
 	meta: string;
 	is_new_reader: string;
 	social_login_connections: string;
 	abtests: string;
 	description: string;
 	user_URL: string;
 	primarySiteSlug: string;
 	localeSlug: string;
 	localeVariant: string;
 	isRTL: string;
 };
```

Finally, are these all supposed to be strings? Even ones like `email_verified`? If they are, and if we're enforcing that, then we don't have to repeat the same thing on each line. It's probably worth doing here anyway for the sake of being explicit and for future diffs that add or change a type, but this is a classic `Record` type…

```ts
export type OptionalUserData = Record<OptionalKeys, string>;

type OptionalKeys =
  | 'display_name'
  | 'username'
  | 'avatar_URL'
  | 'site_count'
  | …
```

Reply:
Thanks I had missed quite a few of the return types, I inspected it in the browser and the email_verified flags etc were in fact booleans. I've updated the types to match what was present at runtime and used that Partial type to save a lot of `x | undefined`, nice 😊.

Reply:
> Is there a particular reason to create this as a `.d.ts` file instead of simply making `user.ts` or `user/types.ts`

So if I call it `types.ts`, other code won't automatically associate the types if I do `import { User } from 'lib/user/user'` etc. In the */packages* we use `types.ts` but also have a `"types": "src/types.ts"` line in the package.json which we can't do here because it's just in `/client`.

I tried naming it `user.ts` but then I got an error when I imported it in another typescript file, I think it didn't know whether to load the typescript or js file. I might be able to play around and get that working but using `user.d.ts` works 😄

Reply:
we can try a couple things but I think having `user.ts` and `user.js` side-by-side is probably asking for disaster 💥 

if `user.d.ts` next to `user.js` works great that's okay

we can also rename `user.js` into `user.ts` directly and add types there. we can do it inline even. now because it's a constructor function that creates the `User` class the typing might be harder.

the third approach addresses this by creating `lib/user/types.ts`, imports that types inside of `lib/user/user.ts`, and "forces" the export into that type with `export default User as User`

there are probably better ways to add types to these constructor-function "classes" but I don't know off hand and I do know it's harder.

Reply:
I tried a number of things to much frustration. I'm not sure if TypeScript supports constructor functions and prototypes the way we use them, especially our singleton instances like this.

Might as well leave it at `user.d.ts` if that works.